File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/eaaddf5db5734cacd0fce60946a15c35.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	p1 = 7
	p2 = 13
)

type Field struct {
	v int
}

func NewField(x int) *Field {
	return &Field{v: x}
}

func (f *Field) pow(a, p int) int {
	r := 1
	e := a
	for p > 0 {
		if p&1 == 1 {
			r = (r * e) % f.v
		}
		e = (e * e) % f.v
		p >>= 1
	}
	return r
}

func (f *Field) inv(a int) int {
	return f.pow(a, f.v-2)
}

type Hash1 struct {
	p Field
	h []int
}

func NewHash1(t string) *Hash1 {
	h := make([]int, len(t))
	cp := p1
	h[0] = int(t[0])
	for i := 1; i < len(t); i++ {
		h[i] = (h[i-1] + int(t[i])*cp) % p1.v
		cp *= p1
	}
	return &Hash1{
		p: p1,
		h: h,
	}
}

type Hash2 struct {
	p Field
	h []int
}

func NewHash2(t string) *Hash2 {
	h := make([]int, len(t))
	cp := p2
	h[0] = int(t[0])
	for i := 1; i < len(t); i++ {
		h[i] = (h[i-1] + int(t[i])*cp) % p2.v
		cp *= p2
	}
	return &Hash2{
		p: p2,
		h: h,
	}
}

type LSegTree struct {
	n         int
	nullValue int
	nullTag   int
	T         []Pair
}

type Pair struct {
	X int
	Y int
}

func NewLSegTree(data []int, nullValue, nullTag int) *LSegTree {
	n := 1 << logceil(len(data))
	T := make([]Pair, 2*n)
	for i := 0; i < len(data); i++ {
		T[n+i].X = data[i]
	}
	for i := n - 1; i > 0; i-- {
		T[i].X = combineValues(T[i<<1].X, T[i<<1|1].X)
	}
	return &LSegTree{
		n:         n,
		nullValue: nullValue,
		nullTag:   nullTag,
		T:         T,
	}
}

func (t *LSegTree) put(x, v int) {
	t.putInner(x, x, v, 1, t.n)
}

func (t *LSegTree) putInner(from, to, v, i, s int) {
	Q := make([]int, 0)
	r := i
	for {
		if from == 0 && to == s-1 {
			applyTag(&T[i].X, v, s)
			combineTag(&T[i].Y, v)
			break
		}
		pushTagDown(i, s)
		s >>= 1
		i <<= 1
		if from&s != 0 {
			from ^= s
			to ^= s
			i ^= 1
		} else if to&s != 0 {
			t.putInner(0, to^s, v, i|1, s)
			to = s - 1
		}
	}

	for i != r {
		T[i>>1].X = combineValues(T[i].X, T[i^1].X)
		i >>= 1
	}
}

func (t *LSegTree) get(x int) int {
	return t.getInner(x, x)
}

func (t *LSegTree) getInner(from, to, i, s int) int {
	for {
		if from == 0 && to == s-1 {
			return T[i].X
		}
		pushTagDown(i, s)
		s >>= 1
		i <<= 1
		if to&s != 0 {
			to ^= s
			if from&s != 0 {
				from ^= s
				i |= 1
			} else {
				a := getInner(from, s-1, i, s)
				b := getInner(0, to, i|1, s)
				return combineValues(a, b)
			}
		}
	}
}

func (t *LSegTree) pushTagDown(i, s int) {
	applyTag(&T[i<<1].X, T[i].Y, s>>1)
	applyTag(&T[i<<1|1].X, T[i].Y, s>>1)
	combineTag(&T[i<<1].Y, T[i].Y)
	combineTag(&T[i<<1|1].Y, T[i].Y)
	T[i].Y = nullTag
}

func combineValues(a, b int) int {
	return a + b
}

func applyTag(a *int, v, s int) {
	*a += v * s
}

func combineTag(a *int, v int) {
	*a += v
}

func logceil(x int) int {
	return 8*32 - __builtin_clz(x)
}

func gcd(a, b int) int {
	if a < b {
		a, b = b, a
	}
	if b != 0 {
		return gcd(b, a%b)
	}
	return a
}

func in(a, b, c int) bool {
	return a <= b && b < c
}

func longestSuffix(x int, pred func(int) bool) int {
	value := nullValue
	return longestSuffixInner2(x, 1, 0, n, value, pred)
}

func longestSuffixInner(i, l, r int, curValue *int, pred func(int) bool) int {
	if l+1 != r {
		pushTagDown(i, r-l)
	}
	nextValue := combineValues(T[i].X, *curValue)
	if pred(nextValue) {
		*curValue = nextValue
		return l
	}
	if l+1 == r {
		return r
	}
	s := longestSuffixInner(i<<1|1, (l+r)/2, r, curValue, pred)
	if s == (l+r)/2 {
		return longestSuffixInner(i<<1, l, (l+r)/2, curValue, pred)
	}
	return s
}

func longestSuffixInner2(x, i, l, r int, curValue *int, pred func(int) bool) int {
	if x >= r-1 {
		return longestSuffixInner(i, l, r, curValue, pred)
	}

	if l+1 == r {
		nextValue := combineValues(T[i].X, *curValue)
		if pred(nextValue) {
			*curValue = nextValue
			return r
		}
		return l
	}

	pushTagDown(i, r-l)
	m := (l+r)/2
	if x < m {
		return longestSuffixInner2(x, i<<1, l, (l+r)/2, curValue, pred)
	}
	s := longestSuffixInner2(x, i<<1|1, (l+r)/2, r, curValue, pred)
	if s == (l+r)/2 {
		return longestSuffixInner(i<<1, l, (l+r)/2, curValue, pred)
	}
	return s
}

func longestPrefix(x int, pred func(int) bool) int {
	value := nullValue
	return longestPrefixInner2(x, 1, 0, n, value, pred) - 1
}

func longestPrefixInner(i, l, r int, curValue *int, pred func(int) bool) int {
	if l+1 != r {
		pushTagDown(i, r-l)
	}
	nextValue := combineValues(T[i].X, *curValue)
	if pred(nextValue) {
		*curValue = nextValue
		return r
	}
	if l+1 == r {
		return l
	}
	s := longestPrefixInner(i<<1, l, (l+r)/2, curValue, pred)
	if s == (l+r)/2 {
		return longestPrefixInner(i<<1|1, (l+r)/2, r, curValue, pred)
	}
	return s
}

func longestPrefixInner2(x, i, l, r int, curValue *int, pred func(int) bool) int {
	if x <= l {
		return longestPrefixInner(i, l, r, curValue, pred)
	}

	if l+1 == r {
		nextValue := combineValues(T[i].X, *curValue)
		if pred(nextValue) {
			*curValue = nextValue
			return r
		}
		return l
	}

	pushTagDown(i, r-l)
	m := (l+r)/2
	if x >= m {
		return longestPrefixInner2(x, i<<1|1, m, r, curValue, pred)
	}
	s := longestPrefixInner2(x, i<<1, l, m, curValue, pred)
	if s == m {
		return longestPrefixInner(i<<1|1, m, r, curValue, pred)
	}
	return s
}

type DDostizhimieStroki struct{}

func (d *DDostizhimieStroki) solve(in *bufio.Reader, out *bufio.Writer) {
	var n int
	fmt.Fscan(in, &n)
	s := make([]byte, n)
	fmt.Fscan(in, &s)
	n = len(s)

	s = append([]byte{'0'}, append(s, '0')...)

	t := make([]byte, 0)
	st := make([]int, len(s))
	for i := 0; i < len(s); i++ {
		if s[i] == '1' && s[i+1] == '1' {
			i++
			continue
		}
		t = append(t, s[i])
		st[i] = len(t) - 1
	}

	p := make([]int, len(s))
	if s[0] == '1' {
		p[0] = 1
	}
	for i := 1; i < len(s); i++ {
		if s[i] == '0' {
			p[i] = 0
		} else {
			p[i] = p[i-1] + 1
		}
	}

	f := make([]int, len(s))
	f[len(s)-1] = 0
	for i := len(s) - 2; i >= 0; i-- {
		if s[i] == '0' {
			f[i] = 0
		} else {
			f[i] = f[i+1] + 1
		}
	}

	h1 := NewHash1(string(t))
	h2 := NewHash2(string(t))

	var q int
	fmt.Fscan(in, &q)

	for i := 0; i < q; i++ {
		var a, b, c int
		fmt.Fscan(in, &a, &b, &c)

		hm1a, hm2a, lena := getH(h1, h2, st, p, f, a, a+c-1)
		hm1b, hm2b, lenb := getH(h1, h2, st, p, f, b, b+c-1)

		if hm1a == hm1b && hm2a == hm2b && lena == lenb {
			fmt.Fprintln(out, "Yes")
		} else {
			fmt.Fprintln(out, "No")
		}
	}

	out.Flush()
}

func getH(h1 *Hash1, h2 *Hash2, st, p, f []int, l, r int) (int, int, int) {
	if l+f[l]-1 >= r {
		return -1, -1, -1
	}

	b := st[l+f[l]]
	e := st[r-p[r]]

	len := e - b + 1

	hm1 := (h1.h[e] - h1.h[b-1]) / h1.p.pow(b)
	if p[r]%2 == 1 {
		len++
		hm1 += h1.p.pow(e - b + 1)
	}

	if f[l]%2 == 1 {
		len++
		hm1 = 1 + hm1*h1.p
	}

	hm2 := (h2.h[e] - h2.h[b-1]) / h2.p.pow(b)
	if p[r]%2 == 1 {
		hm2 += h2.p.pow(e - b + 1)
	}

	if f[l]%2 == 1 {
		hm2 = 1 + hm2*h2.p
	}

	return hm1, hm2, len
}

func main() {
	d := &DDostizhimieStroki{}
	d.solve(bufio.NewReader(os.Stdin), bufio.NewWriter(os.Stdout))
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/C/eaaddf5db5734cacd0fce60946a15c35.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAXN 200200

typedef unsigned int ui;
typedef long long Long;

typedef struct {
    ui v;
} Field;

Field createField(Long x, unsigned int N) {
    Field f;
    f.v = (ui)(x % N);
    return f;
}

Field fieldAdd(Field a, Field b, unsigned int N) {
    if ((Long)a.v + b.v >= N) a.v += b.v - N;
    else a.v += b.v;
    return a;
}

Field fieldSub(Field a, Field b, unsigned int N) {
    if (a.v < b.v) a.v -= b.v - N;
    else a.v -= b.v;
    return a;
}

Field fieldMul(Field a, Field b, unsigned int N) {
    a.v = (ui)((Long)a.v * b.v % N);
    return a;
}

Field fieldPow(Field a, ui p, unsigned int N) {
    Field r = createField(1, N);
    Field e = a;
    while (p) {
        if (p & 1) r = fieldMul(r, e, N);
        e = fieldMul(e, e, N);
        p >>= 1;
    }
    return r;
}

Field fieldInv(Field a, unsigned int N) {
    return fieldPow(a, N - 2, N);
}

Field fieldDiv(Field a, Field b, unsigned int N) {
    return fieldMul(a, fieldInv(b, N), N);
}

Field h1[MAXN], h2[MAXN];
int st[MAXN], p[MAXN], f[MAXN];
char s[MAXN], t[MAXN];

void buildHash(unsigned int p1, unsigned int p2, int tLen) {
    Field cp = createField(p1, 1000000007);
    h1[0] = createField(t[0], 1000000007);
    for (int i = 1; i < tLen; i++) {
        h1[i] = fieldAdd(h1[i - 1], fieldMul(createField(t[i], 1000000007), cp, 1000000007), 1000000007);
        cp = fieldMul(cp, createField(p1, 1000000007), 1000000007);
    }

    cp = createField(p2, 987656927);
    h2[0] = createField(t[0], 987656927);
    for (int i = 1; i < tLen; i++) {
        h2[i] = fieldAdd(h2[i - 1], fieldMul(createField(t[i], 987656927), cp, 987656927), 987656927);
        cp = fieldMul(cp, createField(p2, 987656927), 987656927);
    }
}

void getH(int l, int r, int *len, Field *hm1, Field *hm2) {
    if (l + f[l] - 1 >= r) {
        *len = -1;
        return;
    }

    int b = st[l + f[l]];
    int e = st[r - p[r]];

    *len = e - b + 1;

    *hm1 = fieldDiv(fieldSub(h1[e], h1[b - 1], 1000000007), fieldPow(createField(p1, 1000000007), b, 1000000007), 1000000007);
    if (p[r] % 2 == 1) {
        (*len)++;
        *hm1 = fieldAdd(*hm1, fieldPow(createField(1, 1000000007), e - b + 1, 1000000007), 1000000007);
    }

    if (f[l] % 2 == 1) {
        (*len)++;
        *hm1 = fieldAdd(*hm1, createField(1, 1000000007), 1000000007);
    }

    *hm2 = fieldDiv(fieldSub(h2[e], h2[b - 1], 987656927), fieldPow(createField(p2, 987656927), b, 987656927), 987656927);
    if (p[r] % 2 == 1) {
        *hm2 = fieldAdd(*hm2, fieldPow(createField(1, 987656927), e - b + 1, 987656927), 987656927);
    }

    if (f[l] % 2 == 1) {
        *hm2 = fieldAdd(*hm2, createField(1, 987656927), 987656927);
    }
}

void solve() {
    int n;
    scanf("%d %s", &n, s);
    n = strlen(s);

    s[0] = '0';
    s[n + 1] = '0';
    s[n + 2] = '\0';
    n += 2;

    int tLen = 0;
    for (int i = 0; i < n; i++) {
        if (s[i] == '1' && s[i + 1] == '1') {
            i++;
            continue;
        }
        t[tLen++] = s[i];
        st[i] = tLen - 1;
    }
    t[tLen] = '\0';

    p[0] = (s[0] == '1') ? 1 : 0;
    for (int i = 1; i < n; i++) {
        p[i] = (s[i] == '0') ? 0 : p[i - 1] + 1;
    }

    f[n - 1] = 0;
    for (int i = n - 2; i >= 0; i--) {
        f[i] = (s[i] == '0') ? 0 : f[i + 1] + 1;
    }

    buildHash(7, 13, tLen);

    int q;
    scanf("%d", &q);
    for (int i = 0; i < q; i++) {
        int a, b, c;
        scanf("%d %d %d", &a, &b, &c);
        Field hm1_a, hm1_b, hm2_a, hm2_b;
        int len_a, len_b;

        getH(a, a + c - 1, &len_a, &hm1_a, &hm2_a);
        getH(b, b + c - 1, &len_b, &hm1_b, &hm2_b);

        if (len_a == len_b && hm1_a.v == hm1_b.v && hm2_a.v == hm2_b.v) {
            printf("Yes\n");
        } else {
            printf("No\n");
        }
    }
}

int main() {
    solve();
    return 0;
}
